### PR TITLE
Improve `get_by_key_enumerated` more

### PR DESCRIPTION
**Pull request by JohnTitor**
Follow-up of #86392, this applies the suggestions by @m-ou-se.

r? @m-ou-se

### CODE REVIEW

**Comment by JohnTitor** [comment]
Meanwhile, let's run perf check to see if this has an impact.

@bors try @rust-timer queue

**Comment by m-ou-se** [comment]
This still requires finding the upper bound/amount of items before returning the iterator. That's needed because the key (of type &Q) doesn't live long enough to return it. But since `get_by_key_enumerated` is only ever used by `get_by_key`, and `get_by_key` is only ever used with `K = Symbol` (which is Copy and even smaller than a reference), maybe it's best to remove the `Q` type, and just take a `K` by value. Then you can return the `.take_while()` iterator directly, without .count()'ing it.

**Comment by JohnTitor** [comment]
So, you mean making `SortedIndexMultiMap` `Symbol`-specific? If so, wouldn't it need us to move out `SortedIndexMultiMap` to somewhere?

**Comment by m-ou-se** [comment]
No, I'm suggesting just taking the `K` by value, instead of borrowing a `Q`. Then you can just move it into the closure, and return the .take_while() directly.

**Comment by JohnTitor** [comment]
Ah, that makes sense! Applied the suggestion via d731d61.

File: compiler/rustc_data_structures/src/sorted_map/index_map.rs [hunk 1]
```diff
@@ -89,35 +85,13 @@ impl<I: Idx, K: Ord, V> SortedIndexMultiMap<I, K, V> {
...
-        K: Borrow<Q>,
-    {
-        self.idx_sorted_by_item_key.binary_search_by(|&idx| self.items[idx].0.borrow().cmp(key))
-    }
-
-    fn idxs_to_items_enumerated(&'a self, idxs: &'a [I]) -> impl 'a + Iterator<Item = (I, &'a V)> {
-        idxs.iter().map(move |&idx| (idx, &self.items[idx].1))
+    pub fn get_by_key_enumerated(&'a self, key: K) -> impl '_ + Iterator<Item = (I, &V)> {
+        let lower_bound =
+            self.idx_sorted_by_item_key.partition_point(|&i| self.items[i].0.borrow() < &key);
```
Comment (m-ou-se) [inline compiler/rustc_data_structures/src/sorted_map/index_map.rs:90]: The `.borrow()` was only needed for the `Borrow<Q>`.

File: compiler/rustc_data_structures/src/sorted_map/index_map.rs [hunk 2]
```diff
@@ -89,35 +85,13 @@ impl<I: Idx, K: Ord, V> SortedIndexMultiMap<I, K, V> {
...
+        self.idx_sorted_by_item_key[lower_bound..]
+            .iter()
+            .take_while(move |&&i| self.items[i].0.borrow().eq(&key))
+            .map(move |&idx| (idx, &self.items[idx].1))
```
Comment (m-ou-se) [inline compiler/rustc_data_structures/src/sorted_map/index_map.rs:94]: You can combine these in a single [`map_while`](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.map_while).

**Review by m-ou-se — COMMENTED** [review]
Nice.

(You still need to update the tests.)

Since I'm not on the compiler team or compiler-contributors, assigning it to someone on the compiler team for final approval:

r? @matthewjasper

File: compiler/rustc_data_structures/src/sorted_map/index_map.rs (hunk 2 above)
Comment (JohnTitor) [reply compiler/rustc_data_structures/src/sorted_map/index_map.rs:94]: Hm, IIUC `map_while` does an operation like "do `take_while` while `map` returns `Some`" and this doesn't seem to be a case?

Comment (m-ou-se) [reply compiler/rustc_data_structures/src/sorted_map/index_map.rs]: Yeah you'd have to return a Some if you have an item, or a None if it's done. So then you can check and map at the same time.

E.g. something like

```rust
.take_while(move |&i| {
    let (k, v) = &self.items[i];
    (k == key).then_some((i, v))     // or Some((i, v)).filter(|_| k == key)
})
```

Then you only need to index into `self.items[i]` once.

(But I don't think it really matters after optimization, so pick whatever you find more readable.)

Comment (JohnTitor) [reply compiler/rustc_data_structures/src/sorted_map/index_map.rs]: I see, thanks for the example! I don't have a strong preference but it seems better than accessing by a tuple indexing. Applied your suggestion.

**Pull request merged**